server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndex.java

```diff
@@ -58,36 +62,83 @@ public ComponentIndex(EsClient client, UserSession userSession) {
     this.userSession = userSession;
   }
 
-  public List<String> search(ComponentIndexQuery query) {
+  public List<ComponentsPerQualifier> search(ComponentIndexQuery query) {
+    return search(query, ComponentIndexSearchFeature.values());
+  }
+
+  @VisibleForTesting
+  List<ComponentsPerQualifier> search(ComponentIndexQuery query, ComponentIndexSearchFeature... features) {
+    if (query.getQualifiers().isEmpty()) {
+      return Collections.emptyList();
+    }
+
     SearchRequestBuilder request = getClient()
       .prepareSearch(INDEX_COMPONENTS)
       .setTypes(TYPE_COMPONENT)
-      .setFetchSource(false);
+      .setFetchSource(false)
+
+      // the search hits are part of the aggregations
+      .setSize(0)
+
+      .setQuery(createQuery(query, features));
+
+    query.getQualifiers().stream()
+      .map(q -> createAggregation(query, q))
+      .forEach(request::addAggregation);
+
+    SearchResponse searchResponse = request.get();
+
+    return query.getQualifiers().stream()
+      .flatMap(q -> {
 
-    query.getLimit().ifPresent(request::setSize);
+        InternalFilter agg = searchResponse.getAggregations().get(q);
+        InternalTopHits docs = agg.getAggregations().get("docs");
```

Comment:
This value should be extracted in a constant so that it's shared with definition of aggregation at line 127.